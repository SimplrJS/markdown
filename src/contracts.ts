export interface EmphasisOptions {
    useUnderscores?: boolean;
}

export interface CodeOptions {
    lang?: string;
}

export interface UnorderedListOptions {
    symbol?: UnorderedListSymbols;
}

export type UnorderedListSymbols = "*" | "+" | "-";

export interface TableOptions {
    removeColumnIfEmpty?: boolean;
    removeRowIfEmpty?: boolean;
}

export interface TableHeader {
    text: string;
    align: TableAlign;
}

export type TableAlign = "left" | "center" | "right" | "none";

export interface MarkdownList extends Array<string | MarkdownList> { }

export type HorizontalRuleSymbol = "-" | "*" | "_";

export interface BlockquoteOptions {
    escape: boolean;
    escapeCharacter: string;
}
