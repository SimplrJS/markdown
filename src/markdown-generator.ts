import * as S from "string";

import {
    CodeOptions,
    EmphasisOptions,
    TableHeader,
    TableOptions,
    UnorderedListOptions,
    UnorderedListSymbols,
    MarkdownList,
    HorizontalRuleSymbol
} from "./contracts";

import { ListGenerator } from "./generators/list-generator";
import { TableGenerator } from "./generators/table-generator";

export namespace MarkdownGenerator {
    export function Header(text: string, headerLevel: number, closing: boolean = false): string {
        if (!isFinite(headerLevel)) {
            throw Error("Heading number must be finite!");
        }
        if (headerLevel < 1 || headerLevel > 6) {
            throw Error("Heading number must be more than 0 and less than 7.");
        }

        const headers = S("#").repeat(headerLevel).s;
        const sanitizedText = S(text).trim().s;

        if (closing) {
            return `${headers} ${sanitizedText} ${headers}`;
        } else {
            return `${headers} ${sanitizedText}`;
        }
    }

    export function UnderlineHeader(text: string, headerLevel: number): string[] {
        if (!isFinite(headerLevel)) {
            throw Error("Heading number must be finite!");
        }
        if (headerLevel < 1 || headerLevel > 2) {
            throw Error("Heading number must be more than 0 and less than 2.");
        }

        const sanitizedText = S(text).trim().s;
        const header = headerLevel === 1 ? S("=").repeat(sanitizedText.length).s : S("-").repeat(sanitizedText.length).s;

        return [
            sanitizedText,
            header
        ];
    }

    export function Blockquote(text: string | string[]): string[] {
        let textLines: string[] = [];

        if (typeof text === "string") {
            textLines = S(text).trim().lines();
        } else {
            textLines = text;
        }

        const lines: string[] = [];

        textLines.forEach(x => {
            if (x === "") {
                lines.push(">");
                return;
            }
            /**
             * Trimming only right.
             * Because other markdown elements could have on the left side of the text spacing,
             * for example: Lists.
             */
            const sanitizedText = S(x).trimRight().s;
            lines.push(`> ${sanitizedText}`);
        });

        return lines;
    }

    export function Link(text: string, definitionName: string, definition: boolean): string;
    export function Link(text: string, url: string, linkTitle ?: string): string;
    export function Link(text: string, target: string, arg ?: string | boolean): string {
        const sanitizedText = S(text).trim().s;

        // Link to a link definition
        if (typeof arg === "boolean") {
            return `[${sanitizedText}][${target}]`;
        }

        // Normal link
        let targetText = target;

        if (arg != null) {
            targetText += ` "${arg}"`;
        }

        return `[${sanitizedText}](${targetText})`;
    }

    export function LinkDefinition(text: string, url: string, linkTitle ?: string): string {
        const sanitizedText = S(text).trim().s;
        let linkTitleText = "";

        if (linkTitle != null) {
            linkTitleText = `"${linkTitle}"`;
        }

        return S(`[${sanitizedText}]: ${url} ${linkTitleText}`).trim().s;
    }

    export function Image(altText: string, url: string, title ?: string): string {
        return "!" + Link(altText, url, title);
    }

    export function UnorderedList(list: MarkdownList, options ?: UnorderedListOptions): string[] {
        const allowedSymbols: UnorderedListSymbols[] = [
            "*",
            "+",
            "-"
        ];
        let symbol: UnorderedListSymbols = "*";

        if (options != null && options.symbol != null) {
            if (allowedSymbols.indexOf(options.symbol) === -1) {
                throw Error(`Unordered list: Symbol ${options.symbol} is not allowed. Please use: ${allowedSymbols.join(", ")} instead.`);
            }

            symbol = options.symbol;
        }

        return ListGenerator.RenderList(list, 0, false, symbol);
    }

    export function OrderedList(list: MarkdownList): string[] {
        return ListGenerator.RenderList(list, 0, true);
    }

    export function HorizontalRule(symbol ?: HorizontalRuleSymbol, length ?: number): string {
        const allowedSymbols: HorizontalRuleSymbol[] = [
            "*",
            "-",
            "_"
        ];
        const defaultSymbol: HorizontalRuleSymbol = "-";
        const defaultLength = 3;

        if (symbol != null && allowedSymbols.indexOf(symbol) === -1) {
            throw Error(`Horizontal Rule: Symbol ${symbol} is not allowed. Please use: ${allowedSymbols.join(", ")} instead.`);
        }

        if (length != null && length < 3) {
            throw Error(`Horizontal Rule: Minimum length is ${defaultLength}`);
        }

        const currentSymbol = symbol || defaultSymbol;
        const currentLength = length || defaultLength;

        return S(currentSymbol).pad(currentLength, currentSymbol).s;
    }

    export function Italic(text: string, options ?: EmphasisOptions): string {
        const sanitizedText = S(text).trim().s;

        if (options != null && options.useUnderscores) {
            return `_${sanitizedText}_`;
        }

        return `*${sanitizedText}*`;
    }

    export function Bold(text: string, options ?: EmphasisOptions): string {
        const sanitizedText = S(text).trim().s;

        if (options != null && options.useUnderscores) {
            return `__${sanitizedText}__`;
        }

        return `**${sanitizedText}**`;
    }

    export function StrikeThrough(text: string): string {
        const sanitizedText = S(text).trim().s;

        return `~~${sanitizedText}~~`;
    }

    /**
     * Github flavored markdown
     * @see https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code
     */
    export function InlineCode(text: string): string {
        const sanitizedText = S(text).trim().s;

        return `\`${sanitizedText}\``;
    }

    /**
     * Github flavored markdown
     * @see https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code
     */
    export function Code(text: string | string[], options ?: CodeOptions): string[] {
        let sanitizedText: string[] = [];

        if (typeof text === "string") {
            sanitizedText = S(text).trim().lines();
        } else {
            sanitizedText = text;
        }

        const codeBlockTag = "```";
        let header = codeBlockTag;

        if (options != null && options.lang != null) {
            header += options.lang;
        }

        return [
            header,
            ...sanitizedText,
            codeBlockTag
        ];
    }

    export function Table(headers: string[] | TableHeader[], content: string[][], options ?: TableOptions): string[] {
        return TableGenerator.RenderTable(headers, content, options);
    }

    /**
     * Escapes string.
     * @see https://daringfireball.net/projects/markdown/syntax#backslash
     */
    export function escapeString(text: string): string {
        return text.replace(/[\\\`\*\_\{\}\[\]\(\)\#\+\-\.\!]/g, "\\$&");
    }
}
