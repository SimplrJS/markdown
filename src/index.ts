import * as Contracts from "./contracts";

export * from "./markdown-generator";
export * from "./markdown-builder";
export {
    Contracts
};
