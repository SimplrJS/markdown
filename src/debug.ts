import { MarkdownGenerator } from "./markdown-generator";

function Main(): void {
    const headers: string[] = ["Property", "Value"];
    const rows = [
        ["Name"],
        ["Age"]
    ];

    const result = MarkdownGenerator.Table(headers, rows, { removeColumnIfEmpty: true });
    // tslint:disable-next-line:no-debugger
    debugger;
}

Main();
