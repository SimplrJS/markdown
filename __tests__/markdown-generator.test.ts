import { MarkdownGenerator } from "../src/markdown-generator";
import { TableHeader } from "../src/contracts";
import * as S from "string";

const NEW_LINE = "\n";

describe("Header", () => {
    it("Throw error when number is not finite", () => {
        expect(() => MarkdownGenerator.Header("Some text", "a" as any)).toThrow();
    });

    it("Throw error when number is not valid (0 < x < 7)", () => {
        expect(() => MarkdownGenerator.Header("Some text", 7)).toThrow();
    });

    it("Working example", () => {
        const text = "Some text";
        const result = MarkdownGenerator.Header("Some text", 1);
        expect(result).toMatchSnapshot();
    });

    it("Working example with closing header", () => {
        const text = "Some text";
        const result = MarkdownGenerator.Header(text, 1, true);
        expect(result).toMatchSnapshot();
    });
});

describe("Underline header", () => {
    it("Throw error when number is not finite", () => {
        expect(() => MarkdownGenerator.UnderlineHeader("Some text", "a" as any)).toThrow();
    });

    it("Throw error when number is not valid (0 < x < 3)", () => {
        expect(() => MarkdownGenerator.UnderlineHeader("Some text", 3)).toThrow();
    });

    it("Working example", () => {
        const text = "Some text";
        const result = MarkdownGenerator.UnderlineHeader(text, 1);
        expect(result).toMatchSnapshot();
    });

    it("Working example 2", () => {
        const text = "Some text";
        const result = MarkdownGenerator.UnderlineHeader(text, 2);
        expect(result).toMatchSnapshot();
    });
});

describe("BlockQuotes", () => {
    it("Simple working example", () => {
        const text = "Some text";
        const result = MarkdownGenerator.Blockquote(text);
        expect(result).toMatchSnapshot();
    });

    it("Simple working example with escaping", () => {
        const text = ">Some text";
        const result = MarkdownGenerator.Blockquote(text, { escapeGreaterThanChar: "\\>" });
        expect(result).toMatchSnapshot();
    });

    it("Multiline working example", () => {
        const text = "Some text \n\nOther Text";
        const result = MarkdownGenerator.Blockquote(text);

        expect(result).toMatchSnapshot();
    });

    it("Multiline string array working example", () => {
        const text = [
            "Some text",
            "Other text"
        ];
        const result = MarkdownGenerator.Blockquote(text);

        expect(result).toMatchSnapshot();
    });

    /**
     * @see https://daringfireball.net/projects/markdown/syntax#blockquote
     */
    it("Contains other Markdown elements, including headers, lists, and code blocks", () => {
        let text: string = "";
        const list = [
            "One",
            "Two",
            "Three",
            [
                "Three One",
                "Three Two",
                [
                    "Three Two One"
                ]
            ]
        ];

        const headers: string[] = ["Property", "Value"];
        const rows = [
            ["Name", "string"],
            ["Age", "number"]
        ];

        text += MarkdownGenerator.Table(headers, rows).join(NEW_LINE);
        text += NEW_LINE;
        text += NEW_LINE;
        text += MarkdownGenerator.OrderedList(list).join(NEW_LINE);
        text += NEW_LINE;
        text += NEW_LINE;
        text += "Hello world";

        const result = MarkdownGenerator.Blockquote(text);
        expect(result).toMatchSnapshot();
    });
});

describe("Emphasis", () => {
    it("italic using asterisks", () => {
        const text = "Some text  ";
        const result = MarkdownGenerator.Italic(text);
        expect(result).toMatchSnapshot();
    });

    it("italic using underscores", () => {
        const text = "Some text  ";
        const result = MarkdownGenerator.Italic(text, { useUnderscores: true });
        expect(result).toMatchSnapshot();
    });

    it("bold using asterisks", () => {
        const text = "Some text  ";
        const result = MarkdownGenerator.Bold(text);
        expect(result).toMatchSnapshot();
    });

    it("bold using underscores", () => {
        const text = "Some text  ";
        const result = MarkdownGenerator.Bold(text, { useUnderscores: true });
        expect(result).toMatchSnapshot();
    });

    it("StrikeThrough", () => {
        const text = "Some text  ";
        const result = MarkdownGenerator.StrikeThrough(text);
        expect(result).toMatchSnapshot();
    });
});

describe("Code", () => {
    it("inline", () => {
        const text = "git status";
        const result = MarkdownGenerator.InlineCode(text);
        expect(result).toMatchSnapshot();
    });

    it("inline with escape", () => {
        const text = "git ` status";
        const result = MarkdownGenerator.InlineCode(text);
        expect(result).toMatchSnapshot();
    });

    it("inline without escape", () => {
        const text = "git `hello` status";
        const result = MarkdownGenerator.InlineCode(text, { escapeBacktickChar: undefined });
        expect(result).toMatchSnapshot();
    });

    it("block with array of text without language option", () => {
        const codeExample = [
            "function sum(a: number, b: number): number {",
            "   return a + b;",
            "}"
        ];
        const result = MarkdownGenerator.Code(codeExample);
        expect(result).toMatchSnapshot();
    });

    it("block with array of text without language option with escape", () => {
        const codeExample = [
            "function sum(a: number, b: number): number {",
            "```",
            "   return a + b;",
            "```",
            "}"
        ];
        const result = MarkdownGenerator.Code(codeExample);
        expect(result).toMatchSnapshot();
    });

    it("block with array of text without language option without escape", () => {
        const codeExample = [
            "function sum(a: number, b: number): number {",
            "```",
            "   return a + b;",
            "```",
            "}"
        ];
        const result = MarkdownGenerator.Code(codeExample, { escapeBacktickChar: undefined });
        expect(result).toMatchSnapshot();
    });

    it("block with text without language option", () => {
        const codeExample = "function sum(a: number, b: number): number {\n   return a + b;\n}";
        const result = MarkdownGenerator.Code(codeExample);
        expect(result).toMatchSnapshot();
    });

    it("block with array of text with language option", () => {
        const codeExample = [
            "function sum(a: number, b: number): number {",
            "   return a + b;",
            "}"
        ];
        const result = MarkdownGenerator.Code(codeExample, { lang: "typescript" });
        expect(result).toMatchSnapshot();
    });
});

describe("Lists", () => {
    it("Unordered simple list", () => {
        const list = [
            "One",
            "Two",
            "Three"
        ];

        const result = MarkdownGenerator.UnorderedList(list);
        expect(result).toMatchSnapshot();
    });

    it("Unordered deep list", () => {
        const list = [
            "One",
            "Two",
            "Three",
            [
                "Three One",
                "Three Two",
                [
                    "Three Two One"
                ]
            ]
        ];

        const result = MarkdownGenerator.UnorderedList(list);
        expect(result).toMatchSnapshot();
    });

    it("Unordered multi level list", () => {
        const list = [
            "Level 1",
            "Level 1",
            [
                "Level 2",
                "Level 2",
                [
                    "Level 3",
                    "Level 3",
                    [
                        "Level 4"
                    ],
                    "Level 3",
                    "Level 3"
                ],
                "Level 2",
                "Level 2"
            ],
            "Level 1",
            "Level 1"
        ];

        const result = MarkdownGenerator.UnorderedList(list);
        expect(result).toMatchSnapshot();
    });

    it("Unordered multi level list with empty levels", () => {
        const list = [
            "Level 1",
            "Level 1",
            [],
            [],
            [
                "Level 2",
                [],
                "Level 2",
                [
                    "Level 3",
                    [],
                    "Level 3",
                    [
                        "Level 4"
                    ],
                    "Level 3",
                    [],
                    "Level 3"
                ],
                "Level 2",
                [],
                "Level 2"
            ],
            "Level 1",
            [],
            "Level 1"
        ];

        const result = MarkdownGenerator.UnorderedList(list);
        expect(result).toMatchSnapshot();
    });

    it("Unordered simple list with a custom symbol", () => {
        const list = [
            "One",
            "Two",
            "Three"
        ];

        const result = MarkdownGenerator.UnorderedList(list, {
            symbol: "-"
        });
        expect(result).toMatchSnapshot();
    });

    it("UnorderedList should throw when given not supported symbol", () => {
        expect(() => MarkdownGenerator.UnorderedList([], {
            symbol: "a" as any
        })).toThrow();
    });

    it("Ordered simple list", () => {
        const list = [
            "One",
            "Two",
            "Three"
        ];

        const result = MarkdownGenerator.OrderedList(list);
        expect(result).toMatchSnapshot();
    });

    it("Ordered deep list", () => {
        const list = [
            "One",
            "Two",
            "Three",
            [
                "Three One",
                "Three Two",
                [
                    "Three Two One"
                ]
            ]
        ];

        const result = MarkdownGenerator.OrderedList(list);
        expect(result).toMatchSnapshot();
    });

    it("Ordered multi level list", () => {
        const list = [
            "Level 1 - 1",
            "Level 1 - 2",
            [
                "Level 2 - 1",
                "Level 2 - 2",
                [
                    "Level 3 - 1",
                    "Level 3 - 2",
                    [
                        "Level 4 - 1"
                    ],
                    "Level 3 - 3",
                    "Level 3 - 4"
                ],
                "Level 2 - 3",
                "Level 2 - 4"
            ],
            "Level 1 - 3",
            "Level 1 - 4"
        ];

        const result = MarkdownGenerator.UnorderedList(list);
        expect(result).toMatchSnapshot();
    });

    it("Ordered multi level list with empty levels", () => {
        const list = [
            "Level 1 - 1",
            "Level 1 - 2",
            [],
            [],
            [
                "Level 2 - 1",
                [],
                "Level 2 - 2",
                [
                    "Level 3 - 1",
                    [],
                    "Level 3 - 2",
                    [
                        "Level 4 - 1"
                    ],
                    "Level 3 - 3",
                    [],
                    "Level 3 - 4"
                ],
                "Level 2 - 3",
                [],
                "Level 2 - 4"
            ],
            "Level 1 - 3",
            [],
            "Level 1 - 4"
        ];

        const result = MarkdownGenerator.UnorderedList(list);
        expect(result).toMatchSnapshot();
    });
});

describe("Table", () => {
    it("Simple example", () => {
        const headers: string[] = ["Property", "Value"];
        const rows = [
            ["Name", "string"],
            ["Age", "number"]
        ];
        const result = MarkdownGenerator.Table(headers, rows);
        expect(result).toMatchSnapshot();
    });

    it("Simple example with escape", () => {
        const headers: string[] = ["Property", "Value"];
        const rows = [
            ["Name", "string"],
            ["Age", "number | string"]
        ];
        const result = MarkdownGenerator.Table(headers, rows);
        expect(result).toMatchSnapshot();
    });

    it("Simple example without escape", () => {
        const headers: string[] = ["Property", "Value"];
        const rows = [
            ["Name", "string"],
            ["Age", "number | string"]
        ];
        const result = MarkdownGenerator.Table(headers, rows, { escapePipeChar: undefined });
        expect(result).toMatchSnapshot();
    });

    it("Removes a column if it's empty", () => {
        const headers: string[] = ["Property", "Value"];
        const rows = [
            ["Name"],
            ["Age"]
        ];
        const result = MarkdownGenerator.Table(headers, rows, { removeColumnIfEmpty: true });
        expect(result).toMatchSnapshot();
    });

    it("Table with a single column filled.", () => {
        const headers = ["Name", "Constraint type", "Default type"];
        const rows = [
            ["TValue", "", ""]
        ];

        const result = MarkdownGenerator.Table(headers, rows, { removeColumnIfEmpty: true });
        expect(result).toMatchSnapshot();
    });

    it("Table with an empty column in the middle.", () => {
        const headers = ["Name", "Constraint type", "Default type"];
        const rows = [
            ["TValue", "", "string"],
            ["T", "", "string"],
            ["TKey ", "", "string"],
        ];

        const result = MarkdownGenerator.Table(headers, rows, { removeColumnIfEmpty: true });
        expect(result).toMatchSnapshot();
    });

    it("Table with empty rows", () => {
        const headers = ["Name", "Constraint type", "Default type", "InitialValue"];
        const rows = [
            ["TValue", "", "", "{}"],
            ["", "", "", ""],
            ["TKey", "", "", "{}"],
            ["", "", "", ""]
        ];

        const result = MarkdownGenerator.Table(headers, rows, { removeColumnIfEmpty: true, removeRowIfEmpty: true });
        expect(result).toMatchSnapshot();
    });

    it("Table with rows that has a single cell filled", () => {
        const headers = ["Name", "Constraint type", "Default type", "InitialValue"];
        const rows = [
            ["TValue", "", "", ""],
            ["", "Object", "", ""],
            ["", "", "Object", ""],
            ["", "", "", "{}"]
        ];

        const result = MarkdownGenerator.Table(headers, rows, { removeColumnIfEmpty: true });
    });

    it("Simple example with mixed alignments", () => {
        const headers: Array<TableHeader | string> = [
            {
                text: "Property",
                align: "center"
            },
            "Value"
        ];

        const rows = [
            ["Name", "string"],
            ["Age", "number"]
        ];
        const result = MarkdownGenerator.Table(headers, rows);
        expect(result).toMatchSnapshot();
    });

    it("Headers with all alignments", () => {
        const headers: TableHeader[] = [
            {
                text: "Text1",
                align: "center"
            },
            {
                text: "Text2",
                align: "left"
            },
            {
                text: "Text3",
                align: "none"
            },
            {
                text: "Text4",
                align: "right"
            }
        ];

        const rows = [
            ["One", "Two", "Three", "Four"],
            ["One", "Two", "Three", "Four"]
        ];
        const result = MarkdownGenerator.Table(headers, rows);
        expect(result).toMatchSnapshot();
    });

    it("Headers with missing data in rows", () => {
        const headers: TableHeader[] = [
            {
                text: "Text1",
                align: "center"
            },
            {
                text: "Text2",
                align: "left"
            },
            {
                text: "Text3",
                align: "none"
            },
            {
                text: "Text4",
                align: "right"
            }
        ];

        const rows = [
            ["One", "Two", "Three"],
            ["One"],
            ["One", "Two"]
        ];
        const result = MarkdownGenerator.Table(headers, rows, { removeColumnIfEmpty: true });
        expect(result).toMatchSnapshot();
    });
});

describe("Links", () => {
    it("Simple link", () => {
        const result = MarkdownGenerator.Link("Some text", "https://google.com/");
        expect(result).toMatchSnapshot();
    });

    it("Simple link with a hover text", () => {
        const result = MarkdownGenerator.Link("Some text", "https://google.com/", "Hover text");
        expect(result).toMatchSnapshot();
    });

    it("Link to definition", () => {
        const result = MarkdownGenerator.Link("Some text", "Google", true);
        expect(result).toMatchSnapshot();
    });

    it("Link definition", () => {
        const result = MarkdownGenerator.LinkDefinition("Some text", "https://google.com/");
        expect(result).toMatchSnapshot();
    });

    it("Link definition with a hover text", () => {
        const result = MarkdownGenerator.LinkDefinition("Some text", "https://google.com/", "Hover text");
        expect(result).toMatchSnapshot();
    });
});

describe("Image", () => {
    it("Simple image", () => {
        const result = MarkdownGenerator.Image("Alt text", "https://path.com/image.jpg");
        expect(result).toMatchSnapshot();
    });

    it("Simple image with a hover text", () => {
        const result = MarkdownGenerator.Image("Alt text", "https://google.com/", "Hover text");
        expect(result).toMatchSnapshot();
    });
});

describe("Horizontal rule", () => {
    it("Simple horizontal rule", () => {
        const result = MarkdownGenerator.HorizontalRule();
        expect(result).toMatchSnapshot();
    });

    it("Custom length", () => {
        const result = MarkdownGenerator.HorizontalRule(undefined, 10);
        expect(result).toMatchSnapshot();
    });

    it("Custom symbol", () => {
        const result = MarkdownGenerator.HorizontalRule("*");
        expect(result).toMatchSnapshot();
    });

    it("Throws when length is lower than 3", () => {
        expect(() => MarkdownGenerator.HorizontalRule(undefined, 0)).toThrow();
    });

    it("Throws when symbol is not in allowed list.", () => {
        expect(() => MarkdownGenerator.HorizontalRule("1" as any)).toThrow();
    });
});

describe("Escape string", () => {
    it("Escape url", () => {
        const textToEscape = MarkdownGenerator.Link("Google", "https://google.com/");
        const result = MarkdownGenerator.EscapeString(textToEscape);
        expect(result).toMatchSnapshot();
    });

    it("Escape symbols", () => {
        const textToEscape = "\ ` * _ { } [ ] ( ) # + - . ! | < >";
        const result = MarkdownGenerator.EscapeString(textToEscape);
        expect(result).toMatchSnapshot();
    });

    it("Escaped character in table", () => {
        const headers = ["Name", "Constraint type", "Default type"];
        const rows = [
            ["TValue", "string", "text"],
            ["T", MarkdownGenerator.EscapeString("string | undefined"), "text"],
            ["TKey ", "string", "text"],
            ["TValue", MarkdownGenerator.EscapeString("Map<string>"), "text"]
        ];

        const result = MarkdownGenerator.Table(headers, rows);
        expect(result).toMatchSnapshot();
    });
});
