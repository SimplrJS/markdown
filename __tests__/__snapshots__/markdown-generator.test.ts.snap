// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockQuotes Contains other Markdown elements, including headers, lists, and code blocks 1`] = `
Array [
  "> | Property | Value  |",
  "> | -------- | ------ |",
  "> | Name     | string |",
  "> | Age      | number |",
  ">",
  "> 1. One",
  "> 2. Two",
  "> 3. Three",
  ">     1. Three One",
  ">     2. Three Two",
  ">         1. Three Two One",
  ">",
  "> Hello world",
]
`;

exports[`BlockQuotes Multiline string array working example 1`] = `
Array [
  "> Some text",
  "> Other text",
]
`;

exports[`BlockQuotes Multiline working example 1`] = `
Array [
  "> Some text",
  ">",
  "> Other Text",
]
`;

exports[`BlockQuotes Simple working example 1`] = `
Array [
  "> Some text",
]
`;

exports[`Code block with array of text with language option 1`] = `
Array [
  "\`\`\`typescript",
  "function sum(a: number, b: number): number {",
  "   return a + b;",
  "}",
  "\`\`\`",
]
`;

exports[`Code block with array of text without language option 1`] = `
Array [
  "\`\`\`",
  "function sum(a: number, b: number): number {",
  "   return a + b;",
  "}",
  "\`\`\`",
]
`;

exports[`Code block with text without language option 1`] = `
Array [
  "\`\`\`",
  "function sum(a: number, b: number): number {",
  "   return a + b;",
  "}",
  "\`\`\`",
]
`;

exports[`Code inline 1`] = `"\`git status\`"`;

exports[`Emphasis StrikeThrough 1`] = `"~~Some text~~"`;

exports[`Emphasis bold using asterisks 1`] = `"**Some text**"`;

exports[`Emphasis bold using underscores 1`] = `"__Some text__"`;

exports[`Emphasis italic using asterisks 1`] = `"*Some text*"`;

exports[`Emphasis italic using underscores 1`] = `"_Some text_"`;

exports[`Escape string Escape symbols 1`] = `" \\\\\` \\\\* \\\\_ \\\\{ \\\\} \\\\[ \\\\] \\\\( \\\\) \\\\# \\\\+ \\\\- \\\\. \\\\! &#124;"`;

exports[`Escape string Escape url 1`] = `"\\\\[Google\\\\]\\\\(https://google\\\\.com/\\\\)"`;

exports[`Escape string Escaped character in table 1`] = `
Array [
  "| Name   | Constraint type         | Default type |",
  "| ------ | ----------------------- | ------------ |",
  "| TValue | string                  | text         |",
  "| T      | string &#124; undefined | text         |",
  "| TKey   | string                  | text         |",
]
`;

exports[`Header Working example 1`] = `"# Some text"`;

exports[`Header Working example with closing header 1`] = `"# Some text #"`;

exports[`Horizontal rule Custom length 1`] = `"----------"`;

exports[`Horizontal rule Custom symbol 1`] = `"***"`;

exports[`Horizontal rule Simple horizontal rule 1`] = `"---"`;

exports[`Image Simple image 1`] = `"![Alt text](https://path.com/image.jpg)"`;

exports[`Image Simple image with a hover text 1`] = `"![Alt text](https://google.com/ \\"Hover text\\")"`;

exports[`Links Link definition 1`] = `"[Some text]: https://google.com/"`;

exports[`Links Link definition with a hover text 1`] = `"[Some text]: https://google.com/ \\"Hover text\\""`;

exports[`Links Link to definition 1`] = `"[Some text][Google]"`;

exports[`Links Simple link 1`] = `"[Some text](https://google.com/)"`;

exports[`Links Simple link with a hover text 1`] = `"[Some text](https://google.com/ \\"Hover text\\")"`;

exports[`Lists Ordered deep list 1`] = `
Array [
  "1. One",
  "2. Two",
  "3. Three",
  "    1. Three One",
  "    2. Three Two",
  "        1. Three Two One",
]
`;

exports[`Lists Ordered simple list 1`] = `
Array [
  "1. One",
  "2. Two",
  "3. Three",
]
`;

exports[`Lists Unordered deep list 1`] = `
Array [
  "* One",
  "* Two",
  "* Three",
  "    * Three One",
  "    * Three Two",
  "        * Three Two One",
]
`;

exports[`Lists Unordered simple list 1`] = `
Array [
  "* One",
  "* Two",
  "* Three",
]
`;

exports[`Lists Unordered simple list with a custom symbol 1`] = `
Array [
  "- One",
  "- Two",
  "- Three",
]
`;

exports[`Table Headers with all alignments 1`] = `
Array [
  "| Text1 | Text2 | Text3 | Text4 |",
  "| :---: | :---- | ----- | ----: |",
  "| One   | Two   | Three | Four  |",
  "| One   | Two   | Three | Four  |",
]
`;

exports[`Table Headers with missing data in rows 1`] = `
Array [
  "| Text1 | Text2 | Text3 |",
  "| :---: | :---- | ----- |",
  "| One   | Two   | Three |",
  "| One   |       |       |",
  "| One   | Two   |       |",
]
`;

exports[`Table Removes a column if it's empty 1`] = `
Array [
  "| Property |",
  "| -------- |",
  "| Name     |",
  "| Age      |",
]
`;

exports[`Table Simple example 1`] = `
Array [
  "| Property | Value  |",
  "| -------- | ------ |",
  "| Name     | string |",
  "| Age      | number |",
]
`;

exports[`Table Simple example with mixed alignments 1`] = `
Array [
  "| Property | Value  |",
  "| :------: | ------ |",
  "| Name     | string |",
  "| Age      | number |",
]
`;

exports[`Table Table with a single column filled. 1`] = `
Array [
  "| Name   |",
  "| ------ |",
  "| TValue |",
]
`;

exports[`Table Table with an empty column in the middle. 1`] = `
Array [
  "| Name   | Default type |",
  "| ------ | ------------ |",
  "| TValue | string       |",
  "| T      | string       |",
  "| TKey   | string       |",
]
`;

exports[`Table Table with empty rows 1`] = `
Array [
  "| Name   | InitialValue |",
  "| ------ | ------------ |",
  "| TValue | {}           |",
  "| TKey   | {}           |",
]
`;

exports[`Underline header Working example 1`] = `
Array [
  "Some text",
  "=========",
]
`;

exports[`Underline header Working example 2 1`] = `
Array [
  "Some text",
  "---------",
]
`;
